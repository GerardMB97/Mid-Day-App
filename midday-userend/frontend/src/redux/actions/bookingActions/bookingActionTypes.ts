const bookingActionTypes = {
  SAVE_SELECTION: 'SAVE_SELECTION',
  RESET_BOOKING: 'RESET_BOOKING',
  CREATE_BOOKING: 'CREATE_BOOKING'
};

export default bookingActionTypes;
