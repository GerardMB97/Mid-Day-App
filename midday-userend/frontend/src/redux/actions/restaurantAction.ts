import axios from 'axios';
import restaurantActionTypes from './restaurantActionTypes';

export const loadCategories = () => {
  return async (dispatch: Function) => {
    const categories = await axios.get('http://localhost:5000/api/midday/restaurants/categories');

    dispatch({
      type: restaurantActionTypes.LOAD_CATEGORIES,
      categories
    });
  };
}
;
