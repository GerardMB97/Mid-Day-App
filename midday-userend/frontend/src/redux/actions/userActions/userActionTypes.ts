const userActionTypes = {
  SIGN_UP: 'SIGN_UP',
  SIGN_IN: 'SIGN_IN',
  RESET_STATUS: 'RESET_STATUS'
};

export default userActionTypes;
