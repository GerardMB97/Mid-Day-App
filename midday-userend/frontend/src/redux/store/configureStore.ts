import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import rootReducer from '../reducers';
import initialState from './initialState';

function configureStore () {
  const composeEnhancers = compose;

  return createStore(
    rootReducer,
    initialState,
    composeEnhancers(applyMiddleware(reduxImmutableStateInvariant(), thunk))
  );
}

export default configureStore();
