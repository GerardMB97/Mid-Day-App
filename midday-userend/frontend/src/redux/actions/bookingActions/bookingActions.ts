import bookingActionTypes from './bookingActionTypes';

export const saveMenuSelection = (first, second, dessert, user) => {
  return {
    type: bookingActionTypes.SAVE_SELECTION,
    data: [first, second, dessert],
    user
  };
};

export const resetBooking = () => {
  return {
    type: bookingActionTypes.RESET_BOOKING
  };
};
