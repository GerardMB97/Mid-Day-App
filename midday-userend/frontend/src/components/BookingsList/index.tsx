import { Dispatch, bindActionCreators } from 'redux';
import React from 'react';
import { Text, View, StyleSheet, FlatList, ImageBackground, TouchableWithoutFeedback } from 'react-native';
import { connect } from 'react-redux';
import colors from '../../../colors';
import Icon from 'react-native-vector-icons/Ionicons';

import { deleteBooking, deleteInvitation } from '../../redux/actions/userActions/userActions';
import { getSelectedRestaurant } from '../../redux/actions/restaurantActions/restaurantAction';
import { getBooking } from '../../redux/actions/bookingActions/bookingActions';
import { User } from '../../models';
const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 10
  },
  bookingBadge: {
    width: '100%',
    height: 150,
    shadowOffset: { width: 0, height: 0 },
    shadowColor: 'black',
    shadowRadius: 2,
    shadowOpacity: 1.0,
    position: 'relative',
    marginBottom: 20
  },
  bookingImage: {
    width: '100%',
    height: '100%'
  },
  topContainer: {
    width: '100%',
    height: 50,
    backgroundColor: colors.transparentGray,
    position: 'absolute',
    alignItems: 'center',
    justifyContent: 'center'
  },
  bottomContainer: {
    width: 80,
    height: 50,
    position: 'absolute',
    bottom: 0,
    left: '80%',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around'

  },
  icon: {
    fontSize: 30,
    color: colors.alert
  },
  greenicon: {
    fontSize: 30,
    color: 'green'
  },
  redicon: {
    fontSize: 30,
    color: 'red'
  }
});

function BookingsList ({ user, actions, route, navigation }:{user:User, actions:any, route:any, navigation: any}) {
  const { list }:{list:'invitations' | 'bookings'} = route.params;
  return (
    <View style={styles.container}>
      { user[list].length
        ? <FlatList
      keyExtractor={booking => booking._id}
      data={user[list]}
      renderItem = {({ item }) => {
        const isAdmin = user._id === item.bookingAdmin?._id;
        const company = item.people?.length - 2;
        const companyString = company > 0 ? `os ha invitado a ti y a ${company} m√°s` : 'te ha invitado';
        return (

        <View style={styles.bookingBadge}>
          <ImageBackground
          imageStyle={{ opacity: 0.7 }} style={styles.bookingImage} source={{ uri: item.restaurant?.image }}>

            {list === 'bookings'
              ? <View style={styles.topContainer}><Text>reserva para {item.pax} en {item.restaurant?.name} el {item.date}</Text></View>
              : <View style={styles.topContainer}><Text>{item.bookingAdmin?.name} {companyString} a comer en {item.restaurant?.name} el {item.date}</Text></View>
          }

            {list === 'bookings'
              ? <View style={styles.bottomContainer}>
              {isAdmin &&
                 <TouchableWithoutFeedback onPress={() => actions.deleteBooking(user, item._id)}><Icon style={styles.icon} name="trash-outline"></Icon></TouchableWithoutFeedback>

            }
              <TouchableWithoutFeedback onPress={() => {
                actions.getBooking(item._id);
                actions.getSelectedRestaurant(item.restaurant._id);
                navigation.navigate('RestaurantMenu', { mode: 'edittingBook' });
              }}><Icon style={styles.icon} name="settings-outline"></Icon></TouchableWithoutFeedback>
            </View>
              : <View style={styles.bottomContainer}>
                <TouchableWithoutFeedback onPress={() => {
                  actions.getSelectedRestaurant(item.restaurant._id);
                  actions.getBooking(item._id);
                  navigation.navigate('RestaurantMenu', { mode: 'editting' });
                }}>
                    <Icon style={styles.greenicon} name="restaurant-outline"></Icon></TouchableWithoutFeedback>
                <TouchableWithoutFeedback onPress={() => { actions.deleteInvitation(user._id, item._id, true); }}><Icon style={styles.redicon} name="trash-outline"></Icon></TouchableWithoutFeedback>
              </View>}

          </ImageBackground>
        </View>);
      }

      }
      >

      </FlatList>
        : <Text>No tienes reservas</Text>}
    </View>

  );
}

function mapStateToProps ({ user }:{user:User}) {
  return { user };
}
function mapDispatchToProps (dispatch:Dispatch) {
  return { actions: bindActionCreators({ deleteBooking, deleteInvitation, getSelectedRestaurant, getBooking }, dispatch) };
}
export default connect(mapStateToProps, mapDispatchToProps)(BookingsList);
