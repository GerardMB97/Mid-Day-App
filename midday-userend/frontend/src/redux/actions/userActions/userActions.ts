import axios from 'axios';
import userActionTypes from './userActionTypes';

import { SignUpRoute, SignInRoute, updateUserIsNew, deleteBookingRoute, deleteInvitationRoute, updateBookingPaxRoute, updateAllergiesRoute, newBookingRoute, bookingToRestRoute, bookingToUserRoute } from '../../../constants/dataBase';
import { Dispatch } from 'redux';
import { addInvitation } from '../../../utils';
export const signUp = (name:string, email:string, password:string) => {
  return async (dispatch:Dispatch) => {
    try {
      const { data } = await axios.post(SignUpRoute, { name, email, password });
      dispatch({
        type: userActionTypes.SIGN_UP,
        data
      });
    } catch (error) {
    }
  };
};

export const signIn = (email: string, password: string) => {
  return async (dispatch:Dispatch) => {
    try {
      const { data } = await axios.post(SignInRoute, { email, password });

      dispatch({
        type: userActionTypes.SIGN_IN,
        data
      });
    } catch (error) {
      dispatch({
        type: userActionTypes.SIGN_IN,
        data: null
      });
    }
  };
};

export const updateIsnew = ({ _id }:{_id:string}) => {
  return async (dispatch:Dispatch) => {
    try {
      const { data } = await axios.put(updateUserIsNew, { _id });
      dispatch({
        type: userActionTypes.UPDATE_ISNEW,
        data
      });
    } catch (error) {

    }
  };
};

export const updateUserAllergies = (allergen: string) => {
  return {
    type: userActionTypes.UPDATE_USER_ALLERGIES,
    allergen
  };
};

export const updateAllergiesDB = ({ _id, allergies }:any) => {
  return async (dispatch:Dispatch) => {
    try {
      const { data } = await axios.put(updateAllergiesRoute, { _id, allergies });

      dispatch({
        type: userActionTypes.UPDATE_ALLERGIES_DB,
        user: data
      });
    } catch (error) {

    }
  };
};

export const deleteBooking = ({ email }: {email:string}, bookingId:string) => {
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${deleteBookingRoute}${email}`, { bookingId });

    dispatch({
      type: userActionTypes.DELETE_BOOKING,
      user: data
    });
  };
};

export const deleteInvitation = (userId:string, invitationId:string) => {
  return async (dispatch:Dispatch) => {
    try {
      await axios.put(updateBookingPaxRoute, { userId, bookingId: invitationId });
      const { data } = await axios.put(deleteInvitationRoute, { userId, invitationId });
      dispatch({
        type: userActionTypes.DELETE_INVITATION,
        user: data
      });
    } catch (error) {

    }
  };
};

export const createBooking = (date, hour, bookingAdmin, pax, people, restaurantId) => {
  return async (dispatch:Dispatch) => {
    try {
      const booking = {
        date,
        hour,
        bookingAdmin,
        pax,
        people,
        restaurant: restaurantId

      };
      const { data } = await axios.post(newBookingRoute, booking);

      data.people.forEach(async (person) => { if (bookingAdmin !== person.user) { await addInvitation(person.user, data._id); } });
      const reqBody = {
        bookingId: data._id,
        restaurantId

      };
      await axios.put(bookingToRestRoute, reqBody);
      const userReqBody = {
        bookingId: data._id,
        userId: bookingAdmin
      };
      const user = await axios.put(bookingToUserRoute, userReqBody);
      dispatch({
        type: userActionTypes.CREATE_BOOKING,
        user: user.data
      });
    } catch (error) {

    }
  };
};
