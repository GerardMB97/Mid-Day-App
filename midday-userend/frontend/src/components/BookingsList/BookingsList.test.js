import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { fireEvent, render } from 'react-native-testing-library';
import * as userActions from '../../redux/actions/userActions/userActions';
import * as bookingActions from '../../redux/actions/bookingActions/bookingActions';
import * as restaurantActions from '../../redux/actions/restaurantActions/restaurantAction';
import BookingsList from '.';

jest.mock('react-native-vector-icons/Ionicons', () => 'Icon');
const mockSotre = configureStore();
const user = {
  invitations: [{ _id: '1234', people: [1, 2, 3], restaurant: { _id: '1234' } }],
  bookings: [{ _id: '1234', people: [1, 2], restaurant: { _id: '1234' } }]
};
const store = mockSotre({ user });

describe('Given a BookingsList component', () => {
  beforeEach(() => {
    jest.spyOn(userActions, 'deleteBooking').mockReturnValueOnce({ type: '' });
    jest.spyOn(restaurantActions, 'getSelectedRestaurant').mockReturnValueOnce({ type: '' });
    jest.spyOn(userActions, 'deleteInvitation').mockReturnValueOnce({ type: '' });
    jest.spyOn(bookingActions, 'getBooking').mockReturnValueOnce({ type: '' });
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  describe('When rendered with params.list = invitations', () => {
    test('Then it should match its snapshot', () => {
      const tree = render(<Provider store={store}><BookingsList route={{ params: { list: 'invitations' } }}></BookingsList></Provider>);
      expect(tree).toMatchSnapshot();
    });
  });
  describe('When rendered with params.list = bookings', () => {
    test('Then it should match its snapshot', () => {
      const tree = render(<Provider store={store}><BookingsList route={{ params: { list: 'bookings' } }}></BookingsList></Provider>);
      expect(tree).toMatchSnapshot();
    });
  });
  describe('When button with testID delete is pressed', () => {
    test('Then deleteBookingAction should be invoked', () => {
      const { getByTestId } = render(<Provider store={store}><BookingsList route={{ params: { list: 'bookings' } }}></BookingsList></Provider>);
      const button = getByTestId('delete');
      fireEvent.press(button);
      expect(userActions.deleteBooking).toHaveBeenCalled();
    });
  });
  describe('When button with testID delete is pressed', () => {
    const navigation = {
      navigate: jest.fn()
    };
    test('Then getBooking action should be invoked', () => {
      const { getByTestId } = render(<Provider store={store}><BookingsList navigation ={navigation} route={{ params: { list: 'bookings' } }}></BookingsList></Provider>);
      const button = getByTestId('settings');
      fireEvent.press(button);
      expect(bookingActions.getBooking).toHaveBeenCalled();
    });
  });
  describe('When button with testID deleteInvitation is pressed', () => {
    test('Then deleteInvitation action should be invoked', () => {
      const { getByTestId } = render(<Provider store={store}><BookingsList route={{ params: { list: 'invitations' } }}></BookingsList></Provider>);
      const button = getByTestId('delete invitation');
      fireEvent.press(button);
      expect(userActions.deleteInvitation).toHaveBeenCalled();
    });
  });
  describe('When button with testID edit is pressed', () => {
    const navigation = {
      navigate: jest.fn()
    };
    test('Then deleteIngetSelectedRestaurant action should be invoked', () => {
      const { getByTestId } = render(<Provider store={store}><BookingsList navigation = {navigation} route={{ params: { list: 'invitations' } }}></BookingsList></Provider>);
      const button = getByTestId('edit');
      fireEvent.press(button);
      expect(restaurantActions.getSelectedRestaurant).toHaveBeenCalled();
    });
  });
  describe('When list passed by params has no length', () => {
    const user = {
      invitations: [{ _id: '1234', people: [1, 2, 3], restaurant: { _id: '1234' } }],
      bookings: []
    };
    const store = mockSotre({ user });
    test('Then it should render a text saying no tienes reservas', () => {
      const tree = render(<Provider store={store}><BookingsList route={{ params: { list: 'bookings' } }}></BookingsList></Provider>);
      const noInvitations = tree.getByText('No tienes reservas');
      expect(noInvitations).toBeDefined();
    });
  });
});
