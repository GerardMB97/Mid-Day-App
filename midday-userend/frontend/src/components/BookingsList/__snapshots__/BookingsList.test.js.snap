// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a BookingsList component When rendered with params.list = bookings Then it should match its snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 10,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "_id": "1234",
          "people": Array [
            1,
            2,
          ],
          "restaurant": Object {
            "_id": "1234",
          },
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "height": 150,
              "marginBottom": 20,
              "position": "relative",
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 1,
              "shadowRadius": 2,
              "width": "100%",
            }
          }
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": undefined,
                }
              }
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "height": "100%",
                    "width": "100%",
                  },
                  Object {
                    "opacity": 0.7,
                  },
                ]
              }
            />
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "rgba(242, 242, 242, 0.8)",
                  "height": 50,
                  "justifyContent": "center",
                  "position": "absolute",
                  "width": "100%",
                }
              }
            >
              <Text>
                reserva para 
                 en 
                 el 
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "bottom": 0,
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "space-around",
                  "left": "80%",
                  "position": "absolute",
                  "width": 80,
                }
              }
            >
              <Icon
                accessible={true}
                focusable={true}
                name="trash-outline"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "color": "rgb(247, 199, 48)",
                    "fontSize": 30,
                  }
                }
                testID="delete"
              />
              <Icon
                accessible={true}
                focusable={true}
                name="settings-outline"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "color": "rgb(247, 199, 48)",
                    "fontSize": 30,
                  }
                }
                testID="settings"
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Given a BookingsList component When rendered with params.list = invitations Then it should match its snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 10,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "_id": "1234",
          "people": Array [
            1,
            2,
            3,
          ],
          "restaurant": Object {
            "_id": "1234",
          },
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "height": 150,
              "marginBottom": 20,
              "position": "relative",
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 1,
              "shadowRadius": 2,
              "width": "100%",
            }
          }
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": undefined,
                }
              }
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "height": "100%",
                    "width": "100%",
                  },
                  Object {
                    "opacity": 0.7,
                  },
                ]
              }
            />
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "rgba(242, 242, 242, 0.8)",
                  "height": 50,
                  "justifyContent": "center",
                  "position": "absolute",
                  "width": "100%",
                }
              }
            >
              <Text>
                 
                os ha invitado a ti y a 1 m√°s
                 a comer en 
                 el 
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "bottom": 0,
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "space-around",
                  "left": "80%",
                  "position": "absolute",
                  "width": 80,
                }
              }
            >
              <Icon
                accessible={true}
                focusable={true}
                name="restaurant-outline"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "color": "green",
                    "fontSize": 30,
                  }
                }
                testID="edit"
              />
              <Icon
                accessible={true}
                focusable={true}
                name="trash-outline"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "color": "red",
                    "fontSize": 30,
                  }
                }
                testID="delete invitation"
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
